File: src/formatter.nim
import pathutils

type
  FormatterError* {.pure.} = enum
    Success
    NullPath
    ReadFailed
    MemoryAlloc  ## 仅保留以匹配 C 接口（Nim GC 下不会触发）

  FormatterResult* = object
    error*: FormatterError
    errorMessage*: string

proc processContent*(content: string): string =
  ## 核心格式化逻辑，对应 C 的 process_content()
  ##
  ## 1. 行尾空白（空格/制表符）修剪
  ## 2. 行首空白规范化：
  ##    - 含制表符 → 替换为 4 个空格
  ##    - 只有空格 → 完全移除
  ## 3. 保留原有空行结构

  result = newStringOfCap(content.len)

  var i = 0
  let len = content.len

  while i < len:
    # 硬换行保持原样
    if content[i] == '\n':
      result.add('\n')
      i += 1
      continue

    # 定位行尾
    let lineStart = i
    while i < len and content[i] != '\n':
      i += 1
    let lineEnd = i  # 指向 \n（若存在）或 len

    # 行尾空白修剪
    var trimEnd = lineEnd
    while trimEnd > lineStart and
        (content[trimEnd - 1] == ' ' or content[trimEnd - 1] == '\t'):
      dec trimEnd

    # 扫描行首空白，检测是否存在制表符
    var wp = lineStart
    var hasLeadingTabs = false
    while wp < trimEnd and
        (content[wp] == ' ' or content[wp] == '\t'):
      if content[wp] == '\t':
        hasLeadingTabs = true
      inc wp

    # 输出规范化后的行首空白
    if hasLeadingTabs:
      result.add("    ")  # 4 空格
    # else: 只有空格 → 无输出（完全删除）

    # 输出行内剩余内容（wp ~ trimEnd）
    while wp < trimEnd:
      result.add(content[wp])
      inc wp

    # 若原行以 \n 结尾，输出之
    if i < len and content[i] == '\n':
      result.add('\n')
      inc i

proc formatFile*(path: string): FormatterResult =
  ## 格式化文件
  ##
  ## 流程（对应 C 的 format_file）：
  ## 1. 参数验证 → NullPath
  ## 2. 路径解析
  ## 3. 读文件
  ## 4. processContent 格式化
  ## 5. 写回文件
  ## 6. 返回结果

  # 1. 参数验证
  if path.len == 0:
    result.error = FormatterError.NullPath
    result.errorMessage = "Path parameter is required"
    return

  # 2. 路径解析
  let absolutePath = resolveWorkspacePath(path)
  if absolutePath.len == 0:
    result.error = FormatterError.ReadFailed
    result.errorMessage = "Could not resolve path"
    return

  # 3. 读文件
  var content: string
  try:
    content = readFile(absolutePath)
  except CatchableError as e:
    result.error = FormatterError.ReadFailed
    result.errorMessage = "Error reading file: " & e.msg
    return

  # 4. 格式化
  let processed = processContent(content)

  # 5. 写回
  try:
    writeFile(absolutePath, processed)
  except CatchableError as e:
    result.error = FormatterError.ReadFailed
    result.errorMessage = "Error writing file: " & e.msg
    return

  # 6. 成功
  result.error = FormatterError.Success

File: core/src/formatter.rs
use std::fs;

use crate::pathutils::resolve_workspace_path;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FormatterError {
    Success,
    NullPath,
    ReadFailed,
}

#[derive(Debug, Clone)]
pub struct FormatterResult {
    pub error: FormatterError,
    pub error_message: String,
}

pub fn process_content(content: &str) -> String {
    let mut result = String::with_capacity(content.len());
    let bytes = content.as_bytes();
    let len = bytes.len();
    let mut i = 0;

    while i < len {
        if bytes[i] == b'\n' {
            result.push('\n');
            i += 1;
            continue;
        }

        let line_start = i;
        while i < len && bytes[i] != b'\n' {
            i += 1;
        }
        let line_end = i;

        let mut trim_end = line_end;
        while trim_end > line_start && (bytes[trim_end - 1] == b' ' || bytes[trim_end - 1] == b'\t')
        {
            trim_end -= 1;
        }

        let mut wp = line_start;
        let mut has_leading_tabs = false;
        while wp < trim_end && (bytes[wp] == b' ' || bytes[wp] == b'\t') {
            if bytes[wp] == b'\t' {
                has_leading_tabs = true;
            }
            wp += 1;
        }

        if has_leading_tabs {
            result.push_str("    ");
        }

        while wp < trim_end {
            result.push(bytes[wp] as char);
            wp += 1;
        }

        if i < len && bytes[i] == b'\n' {
            result.push('\n');
            i += 1;
        }
    }

    result
}

pub fn format_file(path: &str) -> FormatterResult {
    if path.is_empty() {
        return FormatterResult {
            error: FormatterError::NullPath,
            error_message: "Path parameter is required".to_string(),
        };
    }

    let absolute_path = resolve_workspace_path(path);
    if absolute_path.is_empty() {
        return FormatterResult {
            error: FormatterError::ReadFailed,
            error_message: "Could not resolve path".to_string(),
        };
    }

    let content = match fs::read_to_string(&absolute_path) {
        Ok(c) => c,
        Err(e) => {
            return FormatterResult {
                error: FormatterError::ReadFailed,
                error_message: format!("Error reading file: {}", e),
            };
        }
    };

    let processed = process_content(&content);

    match fs::write(&absolute_path, &processed) {
        Ok(_) => FormatterResult {
            error: FormatterError::Success,
            error_message: String::new(),
        },
        Err(e) => FormatterResult {
            error: FormatterError::ReadFailed,
            error_message: format!("Error writing file: {}", e),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_process_content_trailing_spaces() {
        let input = "hello   \nworld\t\n";
        let output = process_content(input);
        assert_eq!(output, "hello\nworld\n");
    }

    #[test]
    fn test_process_content_leading_tab_to_spaces() {
        let input = "\tfoo\n\t\tbar\n";
        let output = process_content(input);
        assert_eq!(output, "    foo\n    bar\n");
    }

    #[test]
    fn test_process_content_tabs_and_spaces_mixed_leading() {
        let input = "\t \t hello\n";
        let output = process_content(input);
        assert_eq!(output, "    hello\n");
    }

    #[test]
    fn test_process_content_mixed_whitespace() {
        let input = "  \tfoo\n  bar\n";
        let output = process_content(input);
        assert_eq!(output, "    foo\nbar\n");
    }

    #[test]
    fn test_process_content_spaces_only_leading_removed() {
        let input = "  hello\n    world\n";
        let output = process_content(input);
        assert_eq!(output, "hello\nworld\n");
    }

    #[test]
    fn test_process_content_no_trailing_newline() {
        let input = "hello\nworld";
        let output = process_content(input);
        assert_eq!(output, "hello\nworld");
    }

    #[test]
    fn test_process_content_empty() {
        assert_eq!(process_content(""), "");
    }

    #[test]
    fn test_process_content_empty_lines_preserved() {
        let input = "foo\n\nbar\n";
        let output = process_content(input);
        assert_eq!(output, "foo\n\nbar\n");
    }

    #[test]
    fn test_process_content_no_change_needed() {
        let input = "hello\nworld\n";
        let output = process_content(input);
        assert_eq!(output, "hello\nworld\n");
    }

    #[test]
    fn test_process_content_only_whitespace_line() {
        let input = "   \n\t\n";
        let output = process_content(input);
        assert_eq!(output, "\n\n");
    }

    #[test]
    fn test_format_file_null_path() {
        let result = format_file("");
        assert_eq!(result.error, FormatterError::NullPath);
        assert_eq!(result.error_message, "Path parameter is required");
    }

    #[test]
    fn test_format_file_nonexistent() {
        let result = format_file("/nonexistent_path_12345/file.txt");
        assert_eq!(result.error, FormatterError::ReadFailed);
        assert!(result.error_message.contains("Error reading file"));
    }

    #[test]
    fn test_format_file_success() {
        use std::io::Write;
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_format_success.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        writeln!(f, "  hello   ").unwrap();
        let result = format_file(file_path.to_str().unwrap());
        assert_eq!(result.error, FormatterError::Success);
        let content = std::fs::read_to_string(&file_path).unwrap();
        assert_eq!(content, "hello\n");
        let _ = std::fs::remove_file(&file_path);
    }
}
